### PR TITLE
Batching and less DB intensive descendant queries for availability annotation

### LINKED ISSUE #6602: content import stuck at 100%
### Observed behavior

possibly related: https://github.com/learningequality/kolibri/issues/6502

We have reports of this behavior from the forums:

* [Thread 1443](https://community.learningequality.org/t/import-via-online-error-100-downloaded-but-still-incomplete-error/1443) and [thread 1400](https://community.learningequality.org/t/downloading-task-is-not-completing-its-stuck-at-100/1400)
* [Thread 1456](https://community.learningequality.org/t/cant-sign-in-on-kolibri/1456)

There was also this similar prior report with a few users chiming in, and we thought it had been addressed: [Thread 1396](https://community.learningequality.org/t/import-export-issues/1396/6)

Kolibri not completing

### Expected behavior

Imports work as expected

### User-facing consequences

Cannot import content

### Errors and logs

Example log file: [kolibri.txt](https://github.com/learningequality/kolibri/files/4257287/kolibri.txt)

Some examples of 500 errors people are seeing:

```
{
“request”: {
“path”: “/api/auth/session/”,
“entity”: “”,
“params”: {
“1582667895147”: 1582667895147
},
“headers”: {
“Content-Type”: “application/json”,
“Accept”: “application/json, application/json;q=0.8, text/plain;q=0.5, /;q=0.2”,
“X-CSRFToken”: “WC66F25F0mmnmTMSQH3axSySIbe7jkzM0xUc3K3eCZYCOo84sHzl7PJSKJMHnjwQ”
},
“method”: “POST”,
“canceled”: false
},
“url”: “/api/auth/session/?1582667895147=1582667895147”,
“raw”: {},
“status”: {
“code”: 500,
“text”: “Internal Server Error”
},
“headers”: {
“Content-Length”: “27”,
“Content-Type”: “text/html”,
“Date”: “Tue, 25 Feb 2020 21:59:59 GMT”,
“Server”: “0.0.0.0”,
“Vary”: “Cookie”,
“X-Frame-Options”: “SAMEORIGIN”
},
“entity”: “Server Error (500)”
}
```

```
{
“request”: {
“path”: “/api/device/deviceinfo/”,
“params”: {
“1582203680411”: 1582203680411
},
“headers”: {
“Content-Type”: “application/json”,
“Accept”: “application/json, application/json;q=0.8, text/plain;q=0.5, /;q=0.2”,
“X-CSRFToken”: “jpqvkxHGzPKqHz8NaKpqVjDz4qAXeKpSeWbhyiJn78pjGOuyq97AC2OvdsZw5jEo”
},
“method”: “GET”,
“canceled”: false
},
“url”: “/api/device/deviceinfo/?1582203680411=1582203680411”,
“raw”: {},
“status”: {
“code”: 500,
“text”: “Internal Server Error”
},
“headers”: {
“Content-Length”: “27”,
“Content-Type”: “text/html”,
“Date”: “Thu, 20 Feb 2020 13:03:08 GMT”,
“Server”: “0.0.0.0”,
“Vary”: “Cookie”,
“X-Frame-Options”: “SAMEORIGIN”
},
“entity”: “Server Error (500)”
}
```

```
{
“request”: {
“path”: “/api/device/deviceinfo/”,
“params”: {
“1582203680411”: 1582203680411
},
“headers”: {
“Content-Type”: “application/json”,
“Accept”: “application/json, application/json;q=0.8, text/plain;q=0.5, /;q=0.2”,
“X-CSRFToken”: “jpqvkxHGzPKqHz8NaKpqVjDz4qAXeKpSeWbhyiJn78pjGOuyq97AC2OvdsZw5jEo”
},
“method”: “GET”,
“canceled”: false
},
“url”: “/api/device/deviceinfo/?1582203680411=1582203680411”,
“raw”: {},
“status”: {
“code”: 500,
“text”: “Internal Server Error”
},
“headers”: {
“Content-Length”: “27”,
“Content-Type”: “text/html”,
“Date”: “Thu, 20 Feb 2020 13:03:08 GMT”,
“Server”: “0.0.0.0”,
“Vary”: “Cookie”,
“X-Frame-Options”: “SAMEORIGIN”
},
“entity”: “Server Error (500)”}
```

### Steps to reproduce

…

### Context

At least one user was using  v0.13.1 on Windows 10

### PR DESCRIPTION
**Pull request by rtibbles**
### Summary
0.13 subtly changed the concept of availability to connote user selection of what should be visible.
This required changing how we annotate the availability of content after an update, to preserve this.
This introduced a need to get all currently available nodes and annotate them and their descendants only.
The initial implementation of this resulted in highly intensive DB queries that caused a seemingly unending update request on the DB.
This PR resolves this by:
* copying a strategy from Django MPTT for doing the descendant queries
* batching updates so that larger channels are handled in smaller chunks, and that we are never applying too many constraints at once
* Treating topic nodes and non-topic nodes differently, as non-topic nodes have no descendants, so we do not need the same constraints.

As a bonus, it adds a little extra to the progress tracking so that the progress does not stall at 100% when annotating availability.

### Reviewer guidance
Note - I was able to replicate this in the shell by effectively making a high number of granular selections, this may help for testing this in the interface.

Does content import work as expected?
Is the code readable and the comments clear?

### References
Hopefully fixes #6602

----

### Contributor Checklist

PR process:

- [x] PR has the correct target branch and milestone
- [x] PR has 'needs review' or 'work-in-progress' label
- [x] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

Testing:

- [ ] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [x] Critical and brittle code paths are covered by unit tests

### Reviewer Checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

### CODE REVIEW

File: kolibri/core/content/utils/annotation.py
Comment (indirectlylit) [inline kolibri/core/content/utils/annotation.py]: > for not topics

typo?

**Comment by jonboiser** [comment]
Import with 46 includes/11 excludes went through. Not sure if that pushes the limit, but it did have to batch the annotation into 9 pieces:

```
INFO     Setting availability to True of 2523 LocalFile objects based on passed in checksums
INFO     Setting availability of non-topic ContentNode objects based on LocalFile availability in 9 batches
INFO     Annotating ContentNode objects with children for 5 levels
INFO     Annotating ContentNode objects with children for level 5
INFO     Annotating ContentNode objects with children for level 4
INFO     Annotating ContentNode objects with children for level 3
INFO     Annotating ContentNode objects with children for level 2
INFO     Annotating ContentNode objects with children for level 1
```

Not sure if the Task Metadata code was affected, but this is the first time i've seen a task complete but not hit 100% of the resources:

```
193 of 194 resources (202 MB of 202 MB) imported

```

File: kolibri/core/content/utils/annotation.py
Comment (indirectlylit) [inline kolibri/core/content/utils/annotation.py]: Could move this code into its own `_named_function` to keep things a bit tighter

File: kolibri/core/content/utils/annotation.py
Comment (indirectlylit) [inline kolibri/core/content/utils/annotation.py]: Weird :)

could this be

```python
        return select([ContentNodeTable.c.id]).where(False)
```

or maybe

```python
        return select([ContentNodeTable.c.id]).filter(sqlalchemy.sql.false())
```

File: kolibri/core/content/utils/annotation.py
Comment (indirectlylit) [inline kolibri/core/content/utils/annotation.py]: might help to initialize `i` to the min value?

File: kolibri/core/content/utils/annotation.py
Comment (jonboiser) [inline kolibri/core/content/utils/annotation.py]: maybe put the permalink to this line here instead
```suggestion
    https://github.com/django-mptt/django-mptt/blob/38d46c26ca362c471b097ab96a3616b9b20fb883/mptt/managers.py#L66
```

File: kolibri/core/content/utils/annotation.py
Comment (indirectlylit) [inline kolibri/core/content/utils/annotation.py:230]: Do we need to support these being `None`? The method would be simplified if the signature was just

```python
def set_leaf_nodes_invisible(channel_id, node_ids=[], exclude_node_ids=[]):
    # ...
```

but maybe there's a need for it?

File: kolibri/core/content/utils/annotation.py
Comment (indirectlylit) [inline kolibri/core/content/utils/annotation.py]: Can `max_rght` be less than or equal to `0`?

If not, I think this is the same as

```python
min(CHUNKSIZE, ceil(250 * max_rght / (constraint_count or 1)))
```

File: kolibri/core/content/utils/annotation.py
Comment (indirectlylit) [inline kolibri/core/content/utils/annotation.py]:
```suggestion
        "Removing availability of non-topic ContentNode objects in {} batches of {}".format(
            int(ceil(max_rght / dynamic_chunksize)), dynamic_chunksize
```

File: kolibri/core/content/utils/annotation.py
Comment (indirectlylit) [inline kolibri/core/content/utils/annotation.py]: could pull the contents of this loop out into its own `_named_function`

**Review by indirectlylit — APPROVED** [review]
looks good. I have not tested yet

After merging in I'd also like to cut a beta and get testing from both end users and @mrpau-dev 

Left some non-blocking code style comments. Some of those in `set_leaf_nodes_invisible` would apply to `set_leaf_node_availability_from_local_file_availability` also

File: kolibri/core/content/utils/annotation.py
Comment (rtibbles) [reply kolibri/core/content/utils/annotation.py]: Possibly - I was looking around and couldn't find a definitive answer in docs.

File: kolibri/core/content/utils/annotation.py
Comment (rtibbles) [reply kolibri/core/content/utils/annotation.py]: It's an integer counter, that I am then multiplying by the batch size.

File: kolibri/core/content/utils/annotation.py
Comment (rtibbles) [reply kolibri/core/content/utils/annotation.py:230]: It's `None` because: https://docs.quantifiedcode.com/python-anti-patterns/correctness/mutable_default_value_as_argument.html - but I think you have perhaps just temporarily forgotten this.

File: kolibri/core/content/utils/annotation.py
Comment (rtibbles) [reply kolibri/core/content/utils/annotation.py]: It shouldn't be, no, I was just trying to be systematic and perhaps overly cautious.

**Comment by rtibbles** [comment]
There is a postgres test failure, so this might take a little while longer to fix.

File: kolibri/core/content/utils/annotation.py
Comment (jamalex) [reply kolibri/core/content/utils/annotation.py:230]: He knew it originally, but then the memory was modified as a side effect of another call to the same function! /s

File: kolibri/core/content/utils/annotation.py
Comment (jamalex) [inline kolibri/core/content/utils/annotation.py]: The ordering may not be necessary for our purposes here, and so it may be faster to omit.

File: kolibri/core/content/utils/annotation.py
Comment (jamalex) [inline kolibri/core/content/utils/annotation.py]: I haven't read the rest of the code yet to fully understand how this gets used, but my initial thought: This will include descendants of nodes in between two nodes in node_ids, which aren't in fact descendants of any of the nodes in node_ids. In the referenced [MPTT source](https://github.com/django-mptt/django-mptt/blob/master/mptt/managers.py#L167), it build a composite OR query, rather than joining into one big min/max range.

File: kolibri/core/content/utils/annotation.py
Comment (jamalex) [reply kolibri/core/content/utils/annotation.py]: "non topics"
"non-topic nodes"
"hot topics"?

File: kolibri/core/content/utils/annotation.py
Comment (jamalex) [reply kolibri/core/content/utils/annotation.py]: Ah, I guess perhaps it's needed for the groupby below.

File: kolibri/core/content/utils/annotation.py
Comment (jamalex) [reply kolibri/core/content/utils/annotation.py]: Ah, I'd been interpreting these differently (and you replicate the actual MPTT stuff below). I see that these are instead for the batching purposes. Clarifying that in the comments above may be useful, since you reference this being a port of the MPTT code, but that aspect isn't done by MPTT, so the parallel comparison becomes harder.

File: kolibri/core/content/utils/annotation.py
Comment (jamalex) [reply kolibri/core/content/utils/annotation.py]: https://stackoverflow.com/a/12837029 says you can do something like
`session.query(MyObject).filter(sqlalchemy.sql.false())`

File: kolibri/core/content/utils/annotation.py
Comment (jamalex) [inline kolibri/core/content/utils/annotation.py]: Note that you can also just do things like `SELECT 1 FROM kolibriauth_facilityuser;` (maybe faster...?)

Comment (jamalex) [reply kolibri/core/content/utils/annotation.py]: Example here: https://stackoverflow.com/questions/7646173/sqlalchemy-exists-for-query#comment65330050_13336408

**Review by jamalex — APPROVED** [review]
Whew! Lots going on here! Not sure I wrapped my head around all the details, but will defer to the tests for correctness. Don't think I saw any blockers, mostly just some comments reflecting my confusion that could feed into commenting. Thanks!

**Comment by rtibbles** [comment]
Thanks for the reviews, given the postgres test failures, I will address and integrate the comments prior to merge as a little more work is needed.

File: kolibri/core/content/utils/annotation.py
Comment (indirectlylit) [reply kolibri/core/content/utils/annotation.py:230]: > The first time that the function is called [without an argument], Python creates a persistent list. 

Wow, I had no idea that was the case!

I assume then that another thing I've done in the past has not been doing what I thought:

```python
def func(param=calc_default_value()):
```

File: kolibri/core/content/utils/annotation.py
Comment (rtibbles) [reply kolibri/core/content/utils/annotation.py]: Have tried to separate out the parts ported from MPTT and parts that aren't, so hopefully that will help.

**Pull request merged**
